data-integration/getJRC_GEOPP_DB_csv.py

```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Mar 23 09:00:00 2021
+
+Integrate JRC GEOPP DB into EnerMaps DB
+
+@author: giuseppeperonato
+"""
+
+import json
+import logging
+import os
+import sys
+
+import frictionless
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+from pandas_datapackage_reader import read_datapackage
+
+import utilities
+
+# Constants
+logging.basicConfig(level=logging.INFO)
+DS_ID = 2
+URL = "https://data.jrc.ec.europa.eu/dataset/jrc-10128-10001"
+Force = False  # force integration if (meta)data has not changed
+
+# In Docker
+DB_HOST = os.environ.get("DB_HOST")
+DB_PORT = os.environ.get("DB_PORT")
+DB_USER = os.environ.get("DB_USER")
+DB_PASSWORD = os.environ.get("DB_PASSWORD")
+DB_DB = os.environ.get("DB_DB")
+
+
+def get(url: str, dp: frictionless.package.Package, force: bool = False):
+    """
+
+    Retrieve data and check validity/update.
+
+    Parameters
+    ----------
+    url : str
+        URL to retrieve the data from.
+    dp : frictionless.package.Package
+        Datapackage agains which validating the data.
+    force : Boolean, optional
+        If True, new data will be uploaded even if the same as in the db. The default is False.
+
+    Returns
+    -------
+    DataFrame
+        Data in EnerMaps format.
+    GeoDataFrame
+        Spatial data in EnerMaps format.
+    frictionless.package.Package
+        Pakage descring the data.
+
+    """
+    # Parsing
+    value_vars = ["gross_cap_ele", "ini_cap_ele", "gross_cap_th"]
+    id_vars = ["FID", "fields"]
+    spatial_vars = ["longitude", "latitude"]
+
+    ld = utilities.get_ld_json(url)
+    csv_file = ld["distribution"][0]["contentUrl"]
+    datePublished = ld["datePublished"]
+    name = ld["name"].replace(" ", "_")
+
+    # Inferring and completing metadata
+    logging.info("Creating datapackage for input data")
+    new_dp = frictionless.describe_package(csv_file, stats=True,)  # Add stats
+    # Add date
+    new_dp["datePublished"] = datePublished
+
+    # Add missing valies
+    new_dp.resources[0]["schema"]["missingValues"] = ["NULL"]
+    for field in value_vars:
+        new_dp.resources[0].schema.get_field(field).type = "number"
+
+    # Logic for update
+    if dp != None:
+        # check stats
+        ChangedStats = dp["resources"][0]["stats"] != new_dp["resources"][0]["stats"]
+        ChangedDate = dp["datePublished"] != new_dp["datePublished"]
+
+        if ChangedStats or ChangedDate:
+            logging.info("Data has changed")
+        elif force:
+            logging.info("Forced update")
```

Comment:
It is done by existing continuing executing the code. On the contrary, if no forced update is called, the function will return three `None`s

Reply:
So, if I understand correctly:
- If the data has changed (ChangedStats and/or ChangedDate set to True), you tell that the data has changed, then you make the forced update regardless if force is True or False.
- If the data has not changed and force is False, you tell that the data has not changed and return None
- If the data has not changed and force is True, you tell "forced update" and make the forced update

Do you really want the first case?

Reply:
Yes. The forced update will be mostly use while testing. In production, if the source dataset has changed, then it will be loaded into the db.

Reply:
Ok thanks!
It would be cleaner clearer and however not to depend on prints to control the logic of the code, but to put it directly in the control structures.